### PR TITLE
Permissions checker

### PR DESCRIPTION
**Pull request by carrodher**
Pending:

1.- Check on Windows:
  1.1.- Can we obtain some information?
  1.2.- Manage errors
2.- Check `make`
3.- Tests

### CODE REVIEW

**Comment by carrodher** [comment]
Yesterday I talked with @aruiz14 about use [proto buffers](https://developers.google.com/protocol-buffers/docs/proto3) to create a "tree" to use different permissions. This is a good idea and we will try to implement it in future versions, at this moment I include the `-exclude` options to allow exclude files and directories according to a RegExp

File: cmd/permissions-checker/permissionschecks.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	if Verbose {
+		level = level + "  "
+	}
+
+	for _, f := range files {
+		name := f.Name()
+		relativePath := strings.TrimPrefix(strings.Join([]string{path, name}, "/"), Directory)
+		fullPath := strings.Join([]string{path, name}, "/")
+
+		match, _ := regexp.MatchString(Exclude, relativePath)
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: This could be written as:
```
if match, _ := regexp.MatchString(Exclude, relativePath); match {
    if Verbose {
        fmt.Printf("\x1b[33m%sExcluding %s\n\x1b[0m", level, name)
    }
    continue
}
// ... current else block
mode := f.Mode()
currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
```

File: cmd/permissions-checker/permissionschecks.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			}
+		} else {
+			mode := f.Mode()
+			currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
+			currentGroup, _ := user.LookupGroupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Gid))
+			currentPerm := mode.Perm().String()
+			hasCorrectOwner := checkOwner(currentOwner.Username, DefaultOwner)
+			hasCorrectGroup := checkOwner(currentGroup.Name, DefaultGroup)
+
+			if mode.IsRegular() && (f.Name()[0:1] != "." || ShowHidden) {
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: A few things:
 - You already assigned `f.Name()` to `name`
 - I would early discard hidden files, .e.g. after checking it doesn't match de excluded files, do:
```
if !ShowHidden && strings.HasPrefix(name, ".") {
    continue
}
```
That way you simply ignore them without getting all the properties without showing them
 - Given my previous comment, maybe that flag could be renamed to `ignoreHidden` ? (not a strong opinion though).

File: cmd/permissions-checker/permissionschecks.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ * level - used to print the output in a hierarchical way
+ */
+func FindRecursive(path, level string) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// In verbose mode we are printing the output in a hierarchical way
+	if Verbose {
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: I would increase the level always, not only when the verbose flag is enabled.
Also, given that this only uses spaces I would replace the string with an integer.
You can write a small helper that formats the message and the level in the way you want.
You may also find https://golang.org/pkg/strings/#Repeat useful

File: cmd/permissions-checker/permissionschecks.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			}
+		}
+	}
+}
+
+// checkPermissions return true if the permissions are correct (false in other case)
+/* currentPermissions [string] - permissions that the file or directory has
+ * defaultPermissions [string] - permissions that should have
+ */
+func checkPermissions(currentPermissions, defaultPermissions string) bool {
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: Can you add an example of what permissions are? Is it `755`? How does using `strings.Contains` compare it?
Consider normalizing the user-provided to the expected format before checking

File: cmd/permissions-checker/permissionschecks.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	return strings.Contains(currentPermissions, defaultPermissions)
+}
+
+// checkOwner return true if the owner is correct (false in other case)
+/* currentOwner [string] - owner of the file or directory
+ * defaultOwner [string] - owner that should have the file or directory
+ */
+func checkOwner(currentOwner, defaultOwner string) bool {
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: Is it really needed a helper for this?
In any case, it could be just
```return currentOwner == defaultOwner```
the same applies for the group

File: cmd/permissions-checker/permissionschecks.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ * currentPermissions [string] 	- file/directory current permissions
+ * defaultPermissions [string] 	- file/directory default permissions
+ * currentOwner [string] 				- file/directory current owner
+ * currentGroup [string] 				- file/directory current group
+ * hasCorrectPermissions [bool] - true if the permissions are correct
+ * hasCorrectOwner [bool] 			- true if the owner is correct
+ * hasCorrectGroup [bool] 			- true if the group is correct
+ */
+func printOutput(level, kind, name, fullPath, currentPermissions, defaultPermissions, currentOwner, currentGroup string, hasCorrectPermissions, hasCorrectOwner, hasCorrectGroup bool) {
+	if !Verbose {
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: I would remove the need of providing both `name` and `fullPath` by using `filepath.Base`:
```name := fullPath
if !Verbose {
    name = filepath.Base(fullPath)
}
```
Note the logic is inverted to avoid calculating the base name unnecessarily

File: cmd/permissions-checker/permissionschecks.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	if hasCorrectPermissions && hasCorrectOwner && hasCorrectGroup { // Everything correct
+		if Verbose {
+			fmt.Printf("%s(%s) %s %s %s %s\n", level, kind, name, currentPermissions, currentOwner, currentGroup)
+		}
+	} else if hasCorrectPermissions && (!hasCorrectOwner || !hasCorrectGroup) { // Permissions correct, fails owner or group
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s %s %s (expected %s %s)\n\x1b[0m", level, kind, name, currentPermissions, currentOwner, currentGroup, DefaultOwner, DefaultGroup)
+	} else if !hasCorrectPermissions && (hasCorrectOwner && hasCorrectGroup) { // Permissions wrong, owner and group correct
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s (expected %s) %s %s\n\x1b[0m", level, kind, name, currentPermissions, defaultPermissions, currentOwner, currentGroup)
+	} else if !hasCorrectPermissions && (!hasCorrectOwner || !hasCorrectGroup) { // Nothing correct
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s (expected %s) %s %s (expected %s %s)\n\x1b[0m", level, kind, name, currentPermissions, defaultPermissions, currentOwner, currentGroup, DefaultOwner, DefaultGroup)
```
Comment (aruiz14) [inline cmd/permissions-checker/permissionschecks.go]: If you used a helper you could also have all this color codes in just one place

File: cmd/permissions-checker/main.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	DefaultGroup string
+	// Directory to check
+	Directory string
+)
+
+func main() {
+	var (
+		getVersion bool
+	)
+	flag.StringVar(&Directory, "dir", "/opt/bitnami", "Directory to check")
```
Comment (aruiz14) [inline cmd/permissions-checker/main.go]: Not requiring you to change it but there is no need for all the variables to be global.

File: cmd/permissions-checker/permissionschecks.go (hunk 3 above)
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: Now I am using an `int` and the `repeat` function to convert it into space, about increasing the level only when verbose is enable, I am doing this because if verbose is enable I am printing the output in a hierarchical way, but if not, I am printing the output at the same level, so if I increase the level always, I need to implement a logic in the print function to differentiate when to use the level and when not.

File: cmd/permissions-checker/permissionschecks.go (hunk 4 above)
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: `currentPermissions` is the value obtained from `mode.Perm().String()` and it is something like `-rwxrwxr-x` (note that the first character is `-`)
`defaultPermissions` is the value passed by the user (or the default one) and it can contain the first character `-` or not, that is the reason I am using `strings.Contains`

I can remove the first character in `currentPermissions` to unify it, what do you think?

File: cmd/permissions-checker/main.go (hunk 8 above)
Comment (carrodher) [reply cmd/permissions-checker/main.go]: In this case is used to obtain the relative path

File: cmd/permissions-checker/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+The tool requires a set of parameters to work properly:
+
+- *dir*: Directory to check. Default vale: */opt/bitnami*
+- *dir_default*: Default directory permissions. Default vale: *rwxrwxr-x*
+- *file_default*: Default file permissions. Default vale: *rwrw-r--*
+- *owner*: Default owner. Default vale: *bitnami*
+- *group*: Default group. Default vale: *daemon*
+- *exclude*: Files and/or directories to e excluded.
```
Comment (javsalgar) [inline cmd/permissions-checker/README.md]: to be

File: cmd/permissions-checker/permissionschecks.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ * name [string] 				  			- receive the full path of the file/directory (if verbose only use the name)
+ * currentPermissions [string] 	- file/directory current permissions
+ * defaultPermissions [string] 	- file/directory default permissions
+ * currentOwner [string] 				- file/directory current owner
+ * currentGroup [string] 				- file/directory current group
+ * hasCorrectPermissions [bool] - true if the permissions are correct
+ * hasCorrectOwner [bool] 			- true if the owner is correct
+ * hasCorrectGroup [bool] 			- true if the group is correct
+ */
+func printOutput(level int, kind, name, currentPermissions, defaultPermissions, currentOwner, currentGroup string, hasCorrectPermissions, hasCorrectOwner, hasCorrectGroup bool) {
```
Comment (javsalgar) [inline cmd/permissions-checker/permissionschecks.go]: Maybe a struct would make sense here. Too many parameters

File: cmd/permissions-checker/permissionschecks.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+					hasCorrectPermissions := checkPermissions(currentPerm, DefaultDirPerm)
+					printOutput(level, "d", fullPath, currentPerm, DefaultDirPerm, currentOwner.Username, currentGroup.Name, hasCorrectPermissions, hasCorrectOwner, hasCorrectGroup)
+					FindRecursive(fullPath, level)
+				}
+			}
+		}
+	}
+}
+
+// checkPermissions return true if the permissions are correct (false in other case)
```
Comment (javsalgar) [inline cmd/permissions-checker/permissionschecks.go]: Use the third person: checkPermissions `returns` true... Apply to the rest of the functions

Source: https://blog.golang.org/godoc-documenting-go-code

File: cmd/permissions-checker/main.go (hunk 8 above)
Comment (javsalgar) [reply cmd/permissions-checker/main.go]: Having all those variables as global makes the reading of the other file confusing, as it is referencing them. Maybe passing a Config struct or something similar

File: cmd/permissions-checker/permissionschecks.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s %s %s (expected %s %s)\n\x1b[0m", strings.Repeat(" ", level), kind, name, currentPermissions, currentOwner, currentGroup, DefaultOwner, DefaultGroup)
+	} else if !hasCorrectPermissions && (hasCorrectOwner && hasCorrectGroup) { // Permissions wrong, owner and group correct
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s (expected %s) %s %s\n\x1b[0m", strings.Repeat(" ", level), kind, name, currentPermissions, defaultPermissions, currentOwner, currentGroup)
+	} else if !hasCorrectPermissions && (!hasCorrectOwner || !hasCorrectGroup) { // Nothing correct
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s (expected %s) %s %s (expected %s %s)\n\x1b[0m", strings.Repeat(" ", level), kind, name, currentPermissions, defaultPermissions, currentOwner, currentGroup, DefaultOwner, DefaultGroup)
+	}
+}
+
+// ManageInputs Validate and/or format the parameters entered by the user
+func ManageInputs() {
```
Comment (javsalgar) [inline cmd/permissions-checker/permissionschecks.go]: This would make sense to be in the main.go file, it is not related to the tests. What is more, I am not sure if this refactor would make sense

Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: It has sense because if not when the output is printed you could have lines with double `//`
```
(f) /Applications/wordpress-4.9.4-6/apache2//scripts/ctl.sh -rwxr-xr-x (expected -rw-r--r--) crhernandez admin
```
instead of 
```
(f) /Applications/wordpress-4.9.4-6/apache2/scripts/ctl.sh -rwxr-xr-x (expected -rw-r--r--) crhernandez admin
```

In this way, I unified the criteria regardless of whether the user uses a path ending on `/` or not.

File: cmd/permissions-checker/permissionschecks.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	defaultPermissions    string
+	owner                 string
+	group                 string
+	hasCorrectPermissions bool
+	hasCorrectOwner       bool
+	hasCorrectGroup       bool
+}
+
+// FindRecursive iterates in a directory showing permissions in a recursive way
+/* path [string] 											- where to look for files and directories
```
Comment (javsalgar) [inline cmd/permissions-checker/permissionschecks.go]: I do not know if this is the correct way of documenting parameters. Any expert please confirm.

File: cmd/permissions-checker/main.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+	flag.BoolVar(&getVersion, "version", false, "Show current version")
+	flag.Parse()
+
+	// Unifies the format eliminating the last "/" if exists
+	search.baseDirectory = strings.TrimSuffix(search.baseDirectory, "/")
+	search.exclude = strings.TrimSuffix(search.exclude, "/")
+
+	// Checks if the default permissions introduced by the user are in the Linux format
+	if m, _ := regexp.MatchString("^[-rwx]{9,10}$", defaultPerm.file); !m {
+		log.Fatalf("file_default should be in the Linux format (i.e. \"rw-rw-r--\")\n")
```
Comment (mmikulicic) [inline cmd/permissions-checker/main.go]: no need for the trailing newline in the formats passed to the `log` methods.

consider using simply```
log.Fatalf(`file_default should be in the Linux format (i.e. "rw-rw-r--")`)
```

File: cmd/permissions-checker/main.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		os.Exit(2)
+	} else if m, _ := regexp.MatchString("^[-rwx]{9,10}$", defaultPerm.dir); !m {
```
Comment (mmikulicic) [inline cmd/permissions-checker/main.go]: Prefer:

```
checks := []struct {
    flag string
    value string
}{
  {"file_default", defaultPerm.file},
  {"dir_default", defaultPerm.dir},
}
for _, c := range checks {
  if !unixPermRegexp.MatchString(c.value) {
    log.Fatalf(`%s should be in the Linux format (i.e. "rw-rw-r--")`, c.flag)
  }
}
```
It's not really complicating the code (i.e. it's not adding heavy abstractions) but it protects from copy pasta.

File: cmd/permissions-checker/main.go [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	if getVersion {
+		fmt.Println(VERSION)
+		os.Exit(0)
+	}
+
+	fmt.Printf(`==================================================
+	PERMISSIONS CHECKS
+==================================================
```
Comment (mmikulicic) [inline cmd/permissions-checker/main.go]: I wonder if we could generate this from the flags themselves

File: cmd/permissions-checker/main.go [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+	- Default dir permissions: %s
+	- Owner: %s
+	- Group: %s
+	- Exclude: %s
+	- Include hidden: %t
+	- Verbose: %t
+==================================================
+`, search.baseDirectory, defaultPerm.file, defaultPerm.dir, defaultPerm.owner, defaultPerm.group, search.exclude, search.hidden, verbose)
+
+	fmt.Printf("\x1b[34;1m\n-- Checking permissions --\n\x1b[0m")
```
Comment (mmikulicic) [inline cmd/permissions-checker/main.go]: it wouldn't be a bad idea if these escape codes were extracted to constants, e.g.:

```
const (
	esc   = "\x1b"
	ansiBlue  = esc + "[34;1m"
	ansiReset = esc + "[0m"
)
....
	fmt.Printf("%s\n-- Checking permissions --\n%s\n", ansiBlue, ansiReset)
```

if you want to have just a little bit of fun, perhaps even:

```
type blue string

func (s blue) String() string { return fmt.Sprintf("%s%s%s", ansiBlue, string(s), ansiReset) }

....
	fmt.Printf("%s", blue("\n-- Checking permissions --\n\n"))
...
```

File: cmd/permissions-checker/permissionschecks.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+	"io/ioutil"
+	"log"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"syscall"
+)
+
+type currentPermissions struct {
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: can you document the struct and the fields please?

File: cmd/permissions-checker/permissionschecks.go [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+ * 	defaultPerm.dir [string] 					- default permissions for directories
+ * 	defaultPerm.owner [string] 				- default owner
+ * 	defaultPerm.group [string] 				- default group
+ * search [searchSettings] 						- data regarding search options
+ * 	search.hidden [bool] 							- includes hidden files and directories in the search
+ * 	search.exclude [string] 					- files and/or directories to be excluded (regexp)
+ * 	search.baseDirectory [string]			- base directory
+ * level [int]                        - used to print the output in a hierarchical way
+ */
+func FindRecursive(path string, defaultPerm defaultPermissions, search searchSettings, verbose bool, level int) {
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: have you considered https://golang.org/pkg/path/filepath/#Walk?

File: cmd/permissions-checker/permissionschecks.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+	// In verbose mode, we are printing the output in a hierarchical way
+	if verbose {
+		level++
+	}
+
+	for _, f := range files {
+		name := f.Name()
+
+		relativePath := strings.TrimPrefix(strings.Join([]string{path, name}, "/"), search.baseDirectory)
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: consider computing relative paths with https://golang.org/pkg/path/filepath/#Rel

File: cmd/permissions-checker/permissionschecks.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+		fullPath := strings.Join([]string{path, name}, "/")
+
+		if m, _ := regexp.MatchString(search.exclude, relativePath); m {
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: handle regexp errors.

Probably the search.exclude regexp should be compiled during flag parsing in main, and here we should receive the compiled `*regep.Regexp` structure.

File: cmd/permissions-checker/permissionschecks.go [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+			if verbose {
+				fmt.Printf("\x1b[33m%sExcluding %s\n\x1b[0m", strings.Repeat(" ", level), name)
+			}
+		} else {
+			if search.hidden || !strings.HasPrefix(name, ".") {
+				mode := f.Mode()
+				currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: don't ignore errors

File: cmd/permissions-checker/main.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+type defaultPermissions struct {
+	file  string
+	dir   string
+	owner string
+	group string
+}
+
+type searchSettings struct {
```
Comment (mmikulicic) [inline cmd/permissions-checker/main.go]: please document this structure and the fields. e.g. what is "exclude"?
if it's a regexp, let's make this field be a `*regexp.Regexp` (by compiling the regexp before creating a value of this structure)

File: cmd/permissions-checker/permissionschecks.go [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+
+				currentPerm := currentPermissions{
+					permissions:     mode.Perm().String(),
+					owner:           currentOwner.Username,
+					group:           currentGroup.Name,
+					hasCorrectOwner: currentOwner.Username == defaultPerm.owner,
+					hasCorrectGroup: currentGroup.Name == defaultPerm.group,
+				}
+
+				if mode.IsRegular() {
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: what if it's a symlink, or a fifo, or a setuid file? 
are we just going to ignore it?

File: cmd/permissions-checker/permissionschecks.go [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+			}
+		}
+	}
+}
+
+// checkPermissions returns true if the permissions are correct (false in another case)
+/* currentPermissions [string] - permissions that the file or directory has (i.e. -rwxrwxr-x)
+ * defaultPermissions [string] - permissions that should have (i.e. rwxrwxr-x)
+ */
+func checkPermissions(currentPermissions, defaultPermissions string) bool {
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: how does this style of comments render with `go doc` (or `godoc`)?
just try it locally.

Usually in Go the function documentation describes the parameters in a more "discursive" style. Please just take a look at some examples of the standard library.

File: cmd/permissions-checker/permissionschecks.go [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+	return strings.Contains(currentPermissions, defaultPermissions)
+}
+
+// printOutput Prints the data in different formats according to the situation
+/* level [string] 														- used to print the output in a hierarchical way
+ * kind [string] 															- "f" if it is file or "d" in case of a directory
+ * fullPath [string] 				  								- file/directory full path (if verbose only use the name)
+ * currentPerm [currentPermissions]						- data to print relate
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: this is not the right place to document the fields of these structures: the structures themselves should have comments (as noted above). See standard library documentation: https://golang.org/pkg/net/http/#Cookie

Use `go doc` for a quick commandline tool or  `godoc -http`  for the same web rendering as the stdlib.
So you can see how this doc looks like.

Once stuff lands on github it will be indexed and rendered by https://godoc.org/, so it's a good idea to have good looking docs.

File: cmd/permissions-checker/permissionschecks.go [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+		fullPath = filepath.Base(fullPath)
+	}
+	hierarchy := strings.Repeat(" ", level)
+
+	if currentPerm.hasCorrectPermissions && currentPerm.hasCorrectOwner && currentPerm.hasCorrectGroup { // Everything correct
+		if verbose {
+			fmt.Printf("%s(%s) %s %s %s %s\n", hierarchy, kind, fullPath, currentPerm.permissions, currentPerm.owner, currentPerm.group)
+		}
+	} else if currentPerm.hasCorrectPermissions && (!currentPerm.hasCorrectOwner || !currentPerm.hasCorrectGroup) { // Permissions correct, fails owner or group
+		fmt.Printf("\x1b[31;1m%s(%s) %s %s %s %s (expected %s %s)\n\x1b[0m", hierarchy, kind, fullPath, currentPerm.permissions, currentPerm.owner, currentPerm.group, defaultPerm.owner, defaultPerm.group)
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: as I review these, I see that the small helper for rendering ansi colors I pasted above would be useful to declutter this.

File: cmd/permissions-checker/permissionschecks.go [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+ * 	defaultPerm.owner [string] 								- default owner
+ * 	defaultPerm.group [string] 								- default group
+ */
+func printOutput(level int, kind, fullPath string, currentPerm currentPermissions, defaultPerm defaultPermissions, verbose bool) {
+	if verbose {
+		fullPath = filepath.Base(fullPath)
+	}
+	hierarchy := strings.Repeat(" ", level)
+
+	if currentPerm.hasCorrectPermissions && currentPerm.hasCorrectOwner && currentPerm.hasCorrectGroup { // Everything correct
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: hmm, please factor out common factors:

```
if currentPerm.hasCorrectPermissions {
 if ...
} else {
 if ...
}
```

yes, nested conditionals are hard to follow, but not because of the syntax, because of logic; by flattening it out the logic becomes even harder to follow.

File: cmd/permissions-checker/permissionschecks.go [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+				fmt.Printf("\x1b[33m%sExcluding %s\n\x1b[0m", strings.Repeat(" ", level), name)
+			}
+		} else {
+			if search.hidden || !strings.HasPrefix(name, ".") {
+				mode := f.Mode()
+				currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
+				currentGroup, _ := user.LookupGroupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Gid))
+
+				currentPerm := currentPermissions{
+					permissions:     mode.Perm().String(),
```
Comment (mmikulicic) [inline cmd/permissions-checker/permissionschecks.go]: why are we losing the strong typed https://golang.org/pkg/os/#FileMode type and converting it to string so eagerly here.

Keep the `os.FileMode` type instead of `rwxrwxrwx` strings all over the place would make everything a bit more self-documenting and safe.

File: cmd/permissions-checker/README.md [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+The tool requires a set of parameters to work properly:
+
+- *dir*: Directory to check. Default vale: */opt/bitnami*
+- *dir_default*: Default directory permissions. Default vale: *rwxrwxr-x*
+- *file_default*: Default file permissions. Default vale: *rwrw-r--*
+- *owner*: Default owner. Default vale: *bitnami*
+- *group*: Default group. Default vale: *daemon*
+- *exclude*: Files and/or directories to be excluded (you can use RegExp).
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/README.md]: Yes, the default is `(?!.*)` that doesn't match anything

File: cmd/permissions-checker/README.md [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+
+## Basic usage
+
+The tool is executed as follows:
+
+```
+$> permissions-checker -dir <DIRECTORY> -dir_default <DIR DEFAULT PERM> -file_default <FILE DEFAULT PERM> -owner <DEFAULT OWNER> -group <DEFAULT GROUP> -exclude <REGEXP> -hidden -verbose
+```
+
+The tool requires a set of parameters to work properly:
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/README.md]: All parameters have a default value (the values most used in the cloud images)

File: cmd/permissions-checker/main.go [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+	flag.StringVar(&defaultPerm.group, "group", "daemon", "Default group")
+	flag.StringVar(&search.baseDirectory, "dir", "/opt/bitnami", "Directory to check")
+	flag.StringVar(&search.exclude, "exclude", "(?!.*)", "Files and/or directories to be excluded")
+	flag.BoolVar(&search.hidden, "hidden", false, "Includes hidden files and directories")
+	flag.BoolVar(&verbose, "verbose", false, "Print every file and directory")
+	flag.BoolVar(&getVersion, "version", false, "Show current version")
+	flag.Parse()
+
+	// Unifies the format eliminating the last "/" if exists
+	search.baseDirectory = strings.TrimSuffix(search.baseDirectory, "/")
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/main.go]: The `filepath.Clean()` returned path ends in a slash only if it represents a root directory, such as "/" on Unix. In this case, it is not useful eliminating the final slash, but I used it in a previous step to take advantage of its potential.

File: cmd/permissions-checker/permissionschecks.go (hunk 20 above)
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: Using https://golang.org/pkg/path/filepath/#Walk it is not needed anymore

File: cmd/permissions-checker/permissionschecks.go (hunk 24 above)
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: I just include the feature to print the information regarding "other" types, but just printing, not comparing

**Comment by carrodher** [comment]
@mmikulicic I just implemented suggestions marked with 👍 or commented, I'll work tomorrow on the remaining

cc @javsalgar in case you want take a look to the new changes

File: cmd/permissions-checker/main.go [hunk 33]
```diff
@@ -0,0 +1,100 @@
...
+	flag.StringVar(&defaultPerm.group, "group", "daemon", "Default group")
+	flag.StringVar(&search.baseDirectory, "dir", "/opt/bitnami", "Directory to check")
+	flag.StringVar(&excludeStr, "exclude", "(.?!.*)", "Files and/or directories to be excluded")
+	flag.BoolVar(&search.hidden, "hidden", false, "Includes hidden files and directories")
+	flag.BoolVar(&verbose, "verbose", false, "Print every file and directory")
+	flag.BoolVar(&getVersion, "version", false, "Show current version")
+	flag.Parse()
+
+	// Unifies the format eliminating the last "/" if exists
+	search.baseDirectory = strings.TrimSuffix(filepath.Clean(search.baseDirectory), "/")
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/main.go]: Yes, it is something like standardize, I want to remove the "/" of the end in order to homogenize the paths and dirs.
AFAIK, `filepath.Clean()` doesn't remove it

File: cmd/permissions-checker/main.go [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+	excludeStr = strings.TrimSuffix(filepath.Clean(excludeStr), "/")
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/main.go]: Yes, agree

File: cmd/permissions-checker/main.go [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+
+	search.exclude = regexp.MustCompile(excludeStr)
+
+	// Checks if the default permissions introduced by the user are in the Unix format
+	unixPermRegexp := regexp.MustCompile("^[-rwx]{9}$")
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/main.go]: Yes, I will do it. Yesterday's commits did not correct all the suggestions of your first revision, only those marked with 👍

File: cmd/permissions-checker/permissionschecks.go [hunk 36]
```diff
@@ -0,0 +1,117 @@
...
+		}
+		name := info.Name()
+		if search.exclude.MatchString(path) {
+			if verbose {
+				fmt.Printf(Colorize("yellow", fmt.Sprintf("%sExcluding %s\n", strings.Repeat("  ", level), name)))
+			}
+		} else {
+			if search.hidden || !strings.HasPrefix(name, ".") {
+				mode := info.Mode()
+				currentOwner, _ := user.LookupId(fmt.Sprint(info.Sys().(*syscall.Stat_t).Uid))
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: No, it is on my TODO list. Yesterday's commits did not correct all the suggestions of your first revision, only those marked with 👍

File: cmd/permissions-checker/permissionschecks.go [hunk 37]
```diff
@@ -0,0 +1,117 @@
...
+	level := 0
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		// In verbose mode, we are printing the output in a hierarchical way
+		if verbose {
+			level = strings.Count(path, "/") - strings.Count(dir, "/")
+		}
+		name := info.Name()
+		if search.exclude.MatchString(path) {
+			if verbose {
+				fmt.Printf(Colorize("yellow", fmt.Sprintf("%sExcluding %s\n", strings.Repeat("  ", level), name)))
```
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: The function `Colorize()` returns the original string without change the color

File: cmd/permissions-checker/permissionschecks.go (hunk 36 above)
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/permissionschecks.go]: Done

File: cmd/permissions-checker/main.go (hunk 35 above)
[the remark this answers is not in the document]
Comment (carrodher) [reply cmd/permissions-checker/main.go]: Done

**Comment by carrodher** [comment]
@mmikulicic I just implemented yesterday and today suggestions and answer comments

cc @javsalgar in case you want to take a look at the new changes

**Comment by carrodher** [comment]
This is not going to happen in the short term...
